### PR TITLE
 Optimise layerview extruder visibility options

### PR DESCRIPTION
**Pull request by fieldOfView**
~~This PR removes a typo~~

### CODE REVIEW

**Comment by nallath** [comment]
Hmm, merge conflicts. Could you have a look at them? I think Jack got to the code first...

**Comment by fieldOfView** [comment]
Retracted the removal of duplicate code because of conflicts with recent changes by @jackha

**Comment by fieldOfView** [comment]
This PR refactors the duplicated maximum of 4 checkboxes to turn on/off individual extruders in layerview. It replaces the code that currently has a typo.

This PR also moves to lingering compatibility mode preferences from the general page into the layerview options, and it enables translation of the layerview modes.

File: plugins/LayerView/LayerView.qml [hunk 1]
```diff
@@ -248,42 +242,19 @@ Item
...
-                text: "Extruder 2"
-                visible: !UM.LayerView.compatibilityMode && (UM.LayerView.extruderCount >= 2)
-            }
-            CheckBox {
-                checked: view_settings.extruder2_checked
-                onClicked: {
-                    UM.Preferences.setValue("layerview/extruder2_opacity", checked ? 1.0 : 0.0);
-                }
-                text: "Extruder 3"
-                visible: !UM.LayerView.compatibilityMode && (UM.LayerView.etruderCount >= 3)
```
Comment (fieldOfView) [inline plugins/LayerView/LayerView.qml]: Here's that typo. Byebye

File: plugins/LayerView/LayerView.py [hunk 2]
```diff
@@ -392,10 +389,13 @@ def _updateWithPreferences(self):
...
-        self.setExtruderOpacity(1, float(Preferences.getInstance().getValue("layerview/extruder1_opacity")))
-        self.setExtruderOpacity(2, float(Preferences.getInstance().getValue("layerview/extruder2_opacity")))
-        self.setExtruderOpacity(3, float(Preferences.getInstance().getValue("layerview/extruder3_opacity")))
+        extruder_nr = 0
+        for extruder_opacity in Preferences.getInstance().getValue("layerview/extruder_opacities").split(","):
+            try:
+                opacity = float(extruder_opacity)
+            except ValueError:
+                opacity = 1.0
+            self.setExtruderOpacity(extruder_nr, opacity)
```
Comment (nallath) [inline plugins/LayerView/LayerView.py]: So you always set the extruderOpacity of the first extruder? I'm guessing you missed a extruder_nr += 1 somehwere in the loop.

File: plugins/LayerView/LayerView.qml [hunk 3]
```diff
@@ -214,72 +218,47 @@ Item
...
-            }
-            CheckBox {
-                checked: view_settings.extruder3_checked
-                onClicked: {
-                    UM.Preferences.setValue("layerview/extruder3_opacity", checked ? 1.0 : 0.0);
+            Repeater {
+                model: UM.LayerView.extruderCount
+                CheckBox {
+                    checked: [undefined, ""].indexOf(view_settings.extruder_opacities[index]) >= 0 || view_settings.extruder_opacities[index] > 0.5
```
Comment (nallath) [inline plugins/LayerView/LayerView.qml]: I know what you're doing here, but i don't realy like it. Seems a bit of a weird way of doing this.

Comment (awhiemstra) [inline plugins/LayerView/LayerView.qml]: It is very Pythonic, which is pretty wrong in Javascript code :p

File: plugins/LayerView/LayerView.qml [hunk 4]
```diff
@@ -214,72 +218,47 @@ Item
...
-                visible: !UM.LayerView.compatibilityMode && (UM.LayerView.etruderCount >= 3)
-            }
-            CheckBox {
-                checked: view_settings.extruder3_checked
-                onClicked: {
-                    UM.Preferences.setValue("layerview/extruder3_opacity", checked ? 1.0 : 0.0);
+            Repeater {
+                model: UM.LayerView.extruderCount
+                CheckBox {
+                    checked: view_settings.extruder_opacities[index] > 0.5 || view_settings.extruder_opacities[index] == undefined !! view_settings.extruder_opacities[index] == ""
```
Comment (nallath) [inline plugins/LayerView/LayerView.qml]: I think this should be || ;) 

I can't test the code unfortunately, as I'm forced to use the compatibility mode.

Comment (fieldOfView) [inline plugins/LayerView/LayerView.qml]: Fixed that.

Forced to use the compatibility mode? Then you will be thrilled that I moved the two preferences from the General page to the layerview options.

**Comment by fieldOfView** [comment]
Would be nice if this could get another look. The typo is still there, and this PR replaces the code that the typo is in.

File: plugins/LayerView/LayerView.py [hunk 5]
```diff
@@ -392,10 +389,14 @@ def _updateWithPreferences(self):
 
         self.setLayerViewType(int(float(Preferences.getInstance().getValue("layerview/layer_view_type"))));
 
-        self.setExtruderOpacity(0, float(Preferences.getInstance().getValue("layerview/extruder0_opacity")))
-        self.setExtruderOpacity(1, float(Preferences.getInstance().getValue("layerview/extruder1_opacity")))
-        self.setExtruderOpacity(2, float(Preferences.getInstance().getValue("layerview/extruder2_opacity")))
-        self.setExtruderOpacity(3, float(Preferences.getInstance().getValue("layerview/extruder3_opacity")))
+        extruder_nr = 0
+        for extruder_opacity in Preferences.getInstance().getValue("layerview/extruder_opacities").split(","):
```
Comment (awhiemstra) [inline plugins/LayerView/LayerView.py]: Could we use a different separator? We have run into issues using , as separator before and it is awfully close to localized formatting of floats.

File: plugins/LayerView/LayerView.qml [hunk 6]
```diff
@@ -160,14 +160,18 @@ Item
...
         {
             id: layerViewTypes
-            ListElement {
-                text: "Material Color"
+        }
+
+        Component.onCompleted:
+        {
+            layerViewTypes.append({
+                text: catalog.i18nc("@title:layerview mode", "Material Color"),
```
Comment (awhiemstra) [inline plugins/LayerView/LayerView.qml]: `@label:listbox` for context.

File: plugins/LayerView/LayerView.qml [hunk 7]
```diff
@@ -315,6 +294,22 @@ Item
...
             }
+            CheckBox {
+                checked: view_settings.only_show_top_layers
+                onClicked: {
+                    UM.Preferences.setValue("view/only_show_top_layers", checked ? 1.0 : 0.0);
+                }
+                text: catalog.i18nc("@label", "Only Show Top Layers")
+                visible: UM.LayerView.compatibilityMode
+            }
+            CheckBox {
```
Comment (awhiemstra) [inline plugins/LayerView/LayerView.qml]: Do we even still need this option? It was mostly a workaround for slow top layer generation.

Comment (fieldOfView) [inline plugins/LayerView/LayerView.qml]: For this iteration I did not want to take away functionality.
I have some changes in mind for the draggable rangeslider branch.

File: plugins/LayerView/LayerView.qml (hunk 6 above)
Comment (fieldOfView) [inline plugins/LayerView/LayerView.qml]: Fixed

File: plugins/LayerView/LayerView.py (hunk 5 above)
Comment (fieldOfView) [inline plugins/LayerView/LayerView.py]: Changed to use | as a separator

File: plugins/LayerView/LayerView.qml (hunk 3 above)
Comment (Ghostkeeper) [inline plugins/LayerView/LayerView.qml]: "Pythonic" means a combination of "elegant" and "using correct language syntax for the job". As far as I'm concerned these two attributes can be extended to other languages as well. If it's good for Python it should be good for Javascript, as long as this is also the most elegant syntax for the job.

Comment (awhiemstra) [inline plugins/LayerView/LayerView.qml]: Well, no, to me it means "the way it is done in Python" which is not necessarily the best solution for the job (not even in Python).

File: plugins/LayerView/LayerView.py (hunk 5 above)
Comment (jackha) [inline plugins/LayerView/LayerView.py]: Except you forgot this one ^^. No problem, already merged and fixed locally on my machine

**Comment by jackha** [comment]
Merged into #1479, which will be merged into master in the near future.

**Pull request closed without merging**